inst/fit_model/fit_model.cpp

```diff
@@ -0,0 +1,223 @@
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <vector>
+
+#include <frogger.hpp>
+#include <serialize_eigen.hpp>
+
+template<typename T, int rank>
+Eigen::TensorMap <Eigen::Tensor<T, rank>> tensor_to_tensor_map(Eigen::Tensor <T, rank> &d) {
+  return Eigen::TensorMap < Eigen::Tensor < T, rank >> (d.data(), d.dimensions());
+}
+
+int main(int argc, char *argv[]) {
+  if (argc < 4) {
+    std::cout <<
+              "Usage: fit_model <sim_years> <hiv_steps_per_year> <intput_dir> <output_dir>" <<
+              std::endl;
+    return 1;
+  }
+
+  int sim_years = atoi(argv[1]);
+  int hiv_steps_per_year = atoi(argv[2]);
+  std::string input_dir = argv[3];
+  std::string output_dir = argv[4];
+
+  std::string input_abs = std::filesystem::absolute(input_dir);
+  if (!std::filesystem::exists(input_abs)) {
+    std::cout << "Input dir '" << input_dir << "' does not exist." << std::endl;
+    return 1;
+  }
+
+  std::string output_abs = std::filesystem::absolute(output_dir);
+  if (!std::filesystem::exists(output_abs)) {
+    if (std::filesystem::create_directory(output_abs)) {
+      std::cout << "Created output directory '" << output_abs << "'" << std::endl;
+    } else {
+      std::cout << "Failed to create output directory '" << output_abs << "'" << std::endl;
+    }
+  } else {
+    std::cout << "Writing to existing output directory " << output_abs << "'" << std::endl;
+  }
+
+  if (sim_years > 60) {
+    std::cout << "Running to max no of sim years: 60\n" << std::endl;
+    sim_years = 60;
+  }
+  if (hiv_steps_per_year > 10) {
+    std::cout << "Running max no of HIV steps per years: 10" << std::endl;
+    hiv_steps_per_year = 10;
+  }
+
+  const double dt = (1.0 / hiv_steps_per_year);
+  const int num_genders = 2;
+  const int age_groups_pop = 81;
+  const int fertility_first_age_group = 15;
+  const int age_groups_fert = 35;
+  const int disease_stages = 7;
+  const int treatment_stages = 3;
+  const int hiv_adult_first_age_group = 15;
+  const int adult_incidence_first_age_group = hiv_adult_first_age_group;
+  // Hardcoded 15-49 for now (35 groups within this band)
+  const int pAG_INCIDPOP = 35;
+  const int time_art_start = 30;
+
+  // Set working dir whilst reading files
+  std::filesystem::path old_dir = std::filesystem::current_path();
+  std::filesystem::current_path(input_abs);
+  auto exit_handler = [&old_dir]() {
+    std::filesystem::current_path(old_dir);
+  };
+//  std::atexit(exit_handler);
```

Comment:
This doesn't work here as `atexit` needs a function which returns void, which the lambda doesn't satisfy. Any idea how to get this pattern to work? We don't need it necessarily in this case as we've already converted out output dir into an absolute path so it should always work even if current path has been updated

Reply:
apparently (and empirically) you should not use atexit for this. So either move the directory changing into try/catch blocks or split up your main function?

Reply:
Actually maybe that doesn't matter for now as we write our output using an absolute path which is prepared before this so I think this is ok to just remove this commented out line?